clueless/models.py

```diff
@@ -114,37 +129,123 @@ class Turn(models.Model):
     """
     player = models.ForeignKey(Player)
     game = models.ForeignKey('Game') #Game class not defined yet, referencing by string
+    status = models.IntegerField(choices=STATUS_CHOICES, default=1)
+
+    def getAvailableActions(self):
+        """
+        :return: Set of Action subclass class objects that can be taken at this point
+        """
+        #TODO: Implement this method
+        return(None)
+
+    def takeAction(self, action):
+        """
+        Performs an action
+        :param action: Subclass of Action, which will have its performAction function called
+        :return:
+        """
+        #TODO: implement this method
+        return(None)
+
+    def endTurn(self):
+        """
+        Ends this turn
+        """
+        pass
 
 class Action(models.Model):
     """
     Any player action a player can take during a turn
     """
-    turn = models.ForeignKey(Turn)
+    turn = models.ForeignKey(Turn, blank=True)
+    description = models.CharField(max_length=255, blank=True)
+    status = models.IntegerField(choices=STATUS_CHOICES, default=1)
+
+    def validate(self):
+        """
+        :return: boolean Can the action be validly taken
+        """
+        raise NotImplementedError("Subclasses of Action need to implement abstract method validate")
```

Comment:
It is odd, ESPECIALLY coming from Java, but this is a best practice for implementing abstract methods in Python since there is no abstract keyword or anything like that.
[](https://docs.python.org/3/library/exceptions.html)

_exception NotImplementedError
This exception is derived from RuntimeError. In user defined base classes, abstract methods should raise this exception when they require derived classes to override the method._

  The only other option is to use Abstract Base Classes (abc s, [](https://docs.python.org/3/library/abc.html)) which is an un-intuitive, bulky, confusing, pain in the a** way of implementing fully abstract classes in python.  Wayyyyy overkill for what we are doing.